crates/sui-rpc-benchmark/src/bin/pull_grafana_logs.rs

```diff
@@ -0,0 +1,202 @@
+        "UNKNOWN_NET".to_string()
+    };
+    let substring = env::var("SUBSTRING").unwrap_or_else(|_| "Sampled read request".to_string());
+    let query = format!(
+        r#"{{namespace="{}", container="sui-edge-proxy-mysten"}} |= "{}""#,
+        namespace, substring
+    );
+    debug!("Query: {}", query);
+
+    let now = chrono::Utc::now();
+    let one_day_ago = now - chrono::Duration::days(1);
+    let start = env::var("START").unwrap_or(one_day_ago.format("%Y-%m-%dT%H:%M:%SZ").to_string());
+    let end = env::var("END").unwrap_or(now.format("%Y-%m-%dT%H:%M:%SZ").to_string());
+    let limit: Option<u64> = env::var("LIMIT").ok().and_then(|l| l.parse().ok());
+    let client = reqwest::Client::new();
+
+    let mut all_logs = Vec::new();
+    let mut offset = None;
+    loop {
+        let chunk_limit = match limit {
+            Some(l) => {
+                let fetched = all_logs.len() as u64;
+                if fetched >= l {
+                    break;
+                }
+                std::cmp::min(MAX_LOGS_PER_REQUEST, l - fetched)
+            }
+            None => MAX_LOGS_PER_REQUEST,
+        };
+
+        let response = fetch_logs(
+            &client,
+            &grafana_url,
+            &query,
+            &start,
+            &end,
+            chunk_limit,
+            offset,
+        )
+        .await?;
+        let batch: Vec<_> = response
+            .data
+            .result
+            .into_iter()
+            .flat_map(|result| {
+                result
+                    .values
+                    .into_iter()
+                    .map(|(_, message)| GrafanaLog { message })
+            })
+            .collect();
+        // If we have no logs, break
+        if batch.is_empty() {
+            break;
+        }
+
+        let batch_len = batch.len();
+        all_logs.extend(batch);
+        offset = Some(offset.unwrap_or(0) + batch_len as u64);
+    }
+
+    info!("Found {} logs.", all_logs.len());
+
+    let mut method_map: HashMap<String, Vec<String>> = HashMap::new();
+    for log_entry in all_logs {
+        if let Some(body_content) = extract_body_from_message(&log_entry.message) {
+            if let Ok(parsed) = serde_json::from_str::<Value>(&body_content) {
+                let method = parsed
+                    .get("method")
+                    .and_then(|m| m.as_str())
+                    .unwrap_or("unknown_method")
+                    .to_string();
+                method_map.entry(method).or_default().push(body_content);
+            }
+        }
+    }
+
+    let file = File::create("sampled_read_requests.jsonl")?;
+    let mut writer = BufWriter::new(file);
+
+    for (method, bodies) in method_map {
+        info!("Writing {} logs for method: {}", bodies.len(), method);
+        for body in bodies {
+            let line = format!(r#"{{"method":"{}", "body":{}}}"#, method, body);
+            writer.write_all(line.as_bytes())?;
+            writer.write_all(b"\n")?;
```

Comment:
There's no need to allocate for the line before writing it out.

```suggestion
            write!(writer, r#"{{"method":"{method}", "body":"#)?;
            writer.write_all(body.as_bytes())?;
            write!(writer, "}}\n")?;
```